### PR TITLE
MB-16104: v2 prime api for new fields

### PR DESCRIPTION
**Pull request by ronaktruss**
## Summary

Initially, this followed the instructions in this [ADR](https://transcom.github.io/mymove-docs/docs/adrs/api-versioning/) and implements the changes to support a singular new API end point `prime/v2/move-task-orders/{moveTaskOrderID}`. I followed these three example PRs that were linked in the ADR:

- [For yaml changes](https://github.com/transcom/mymove/pull/10816/commits/0f387d655e098e589b10ab27b388bc771e873f09)
- [For handler changes](https://github.com/transcom/mymove/pull/10816/commits/0c4be3bab159a49f83e407830a9fecb9d27278da)
- [For service changes](https://github.com/transcom/mymove/compare/0c4be3b...8ad4aed)

If you are trying to get a sense of what making these changes was like, please take a look at the individual commits. 

### Yaml changes

Making these changes was quite cumbersome. For a larger entity like move_task_order we need a lot of the existing definitions. This meant a lot of manual copy pasting. This also made it difficult to keep the changes in line with main since the original prime.yaml was actively being modified. 

### Handler changes

This was fairly straight-forward yet cumbersome. Adding the model to payload functions for such a large entity meant we were primarily copying functions over with minor modifications. 

### Service changes

Outside of the initial flag setup, this was the most smooth portion of the changes. We had a service for the v1 API and a service for the v2 API. If you diff the two files you can see what the differences are between the two. We could potentially make some enhancements on how we are setting these flags and add some additional logic to make the behavior more consistent between when API calls are being made and when tests are being run. Right now the flag that we are using is nil when we are running tests. (There are some comments about this [[1](https://github.com/transcom/mymove/pull/10952#discussion_r1253262344) | [2](https://github.com/transcom/mymove/pull/10952#discussion_r1253264179)].)

### Tests

The tests are in many ways just duplication of existing tests with minor edits. It's important to call out that the approach that we are taking here make it so that the default behavior utilizes the newer v2 service. This means all existing tests will point to the new v2 service.

### Yaml

I ended up moving some of the existing definitions into their own files so that they can be used across yaml files. These changes are done in #11042. Currently this PR is set to merge into that branch so that we could better isolate these different changes. This will "theoretically" cut down on the duplicated code and make it easier to maintain. Imo, it also makes it a bit easier to identify what structures have changed from v1 to v2. For example in this case its the MoveTaskOrder and Order objects. For viewing purposes a developer can still reference the generated swagger files or they can use [go-swagger to expand/flatten](https://goswagger.io/use/transform.html) to see all of the definitions pulled into a single file. 

### Services/Flags

Previously in order to make the service changes we had introduced some additional flags in the appContext to determine the version of the API that was being called. This obfuscated complexity by making the flag setup happen "behind the scenes" and was something that had impacts on how tests were running without being very explicit. I ended up removing the flags and going with a service approach that was similar to [option 1 ](https://github.com/transcom/mymove/pull/10816/commits/0913f2f4625450a2ac43b358e1890d17d77b747d)where we would just introduce a new service as needed. For this specific instance, I ended up just overriding the existing services as the change was fairly minimal and didn't have significant downsides. If we were introducing say 5 new eager loaded fields that would not always be needed we could have created a separate function to do this and created some reusable local functions that abstracted away some of the shared logic. 

## Other Callouts

During this implementation there were a handful of changes to the model to payload file. Since we had to basically duplicate and adapt those existing functions for the v2 generated files this means that this is an area where there might be errors due to manual input. When reviewing this PR pay special attention to the model to payload functions to ensure that the new version matches the existing versions for all fields that haven't changed.

We do expect this approach to evolve and adapt over time. The purpose of this change is to introduce some plumbing and foundation for future changes so that folks have an example they can refer to if they need direction on how to structure their changes.

If this is merged, we are planning to update the ADR to reflect the changes made here and to potentially add some additional documentation documenting the findings we discovered during this process.

### Considerations for the future

We may want to add some codeowners to the generated swagger files so that there is a manual process for accepting changes that may impact APIs.

We also probably want to spend some time spreading awareness around how modifying existing services can and may impact existing APIs and what the implications of those changes are when we have consumers of our APIs.

It may be worthwhile to explore some other libraries or writing an external wrapper that enables us to prevent the duplication of code around the generated files and the model to payload conversions when the objects are identical.

## Verification Steps for the Reviewer

Take a look at the `Other Callouts` section.

1. Verify the new API is working correctly
2. Verify the old API is still working as expected
3. Verify the new prime_v2.yaml file and corresponding docs
4. Verify the existing prime.yaml file and corresponding docs

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/run-storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

For folks that have made some changes to the payload to model file recently, if you can test scenarios that have those newly added/modified fields populated that would be ideal.

1. Compare and contrast the responses between the v1 and v2 API where :moveTaskOrderID is a path variable
    a. In postman make a get call to the v1 API `https://primelocal:9443/prime/v1/move-task-orders/:moveTaskOrderID`
    b. In postman make a get call to the v2 API `https://primelocal:9443/prime/v2/move-task-orders/:moveTaskOrderID`
    
    You should see they are identical except in the v2 version the fields `methodOfPayment`, `naics`, `packingAndShippingInstructions`,  and `supplyAndServicesCostEstimate` are populated within the orders object and the `contractNumber` is populated within the top level object.
    
    image

2. You can also compare the v1 API response on this branch with main to ensure that the v1 branch has not changed.
3. To test the docs locally you can close the backend server since it runs on port 8080 and then run `openapi preview-docs swagger/prime_v2.yaml`. You can then visit `localhost:8080` and you should be able compare and contrast the v1 API docs with the v2 docs. Ensure that the new fields exist on the v2 doc. 
4. If you want to ensure that the v1 API hasn't changed you can ensure that the prime.yaml in the swagger folder hasn't been modified.

### CODE REVIEW

File: pkg/handlers/config.go [hunk 1]
```diff
@@ -139,6 +140,13 @@ func (c *Config) AuditableAppContextFromRequestWithErrors(
 	var resp middleware.Responder
 	appCtx := c.AppContextFromRequest(r)
 	err := appCtx.NewTransaction(func(txnAppCtx appcontext.AppContext) error {
+		// There may be a better place to do this. It probably should be in its own file.
+		// apiVersion := apiversion.NoneSpecified
+		if r.URL != nil {
+			apiVersion := apiversion.DetermineAPIVersion(r.URL.Path)
+			txnAppCtx.SetAPIVersion(apiVersion)
+		}
```
Comment (ronaktruss) [inline pkg/handlers/config.go:148]: I had to add an if check for a certain tests otherwise we would get a null pointer error. 
Do we want to the default value to be NoneSpecified or would we prefer it be nil when we don't get there?

File: pkg/services/move_task_order/move_task_order_fetcher.go [hunk 2]
```diff
@@ -93,136 +91,16 @@ func (f moveTaskOrderFetcher) ListAllMoveTaskOrders(appCtx appcontext.AppContext
...
-
-	// Filtering external vendor shipments in code since we can't do it easily in Pop without a raw query.
-	// Also, due to a Pop bug, we cannot EagerPreload "Reweigh" or "PPMShipment" likely because they are both
-	// a pointer and "has_one" field, so we're loading those here.  This seems similar to other EagerPreload
-	// issues we've found (and sometimes fixed): https://github.com/gobuffalo/pop/issues?q=author%3Areggieriser
-	var filteredShipments models.MTOShipments
-	if mto.MTOShipments != nil {
-		filteredShipments = models.MTOShipments{}
+	if apiVersion != nil && *apiVersion == apiversion.PrimeVersion2 {
+		return movetaskorderv2.FetchMoveTaskOrder(appCtx, searchParams)
```
Comment (ronaktruss) [inline pkg/services/move_task_order/move_task_order_fetcher.go:101]: We need to check if the apiVersion is nil for tests because there are scenarios where this is nil.

File: pkg/services/move_task_order/move_task_order_fetcher_test.go [hunk 3]
```diff
@@ -288,6 +290,80 @@ func (suite *MoveTaskOrderServiceSuite) TestMoveTaskOrderFetcher() {
 		_, err := mtoFetcher.FetchMoveTaskOrder(suite.AppContextForTest(), &searchParams)
 		suite.Error(err)
 	})
+
+	suite.Run("Compares v2 with v1", func() {
```
Comment (ronaktruss) [inline pkg/services/move_task_order/move_task_order_fetcher_test.go:294]: Added an explicit test to test v1 since it would otherwise not be called and we would have no tests verifying the previous behavior.

File: pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	"github.com/transcom/mymove/pkg/services"
+	"github.com/transcom/mymove/pkg/services/move_task_order/shared"
+)
+
+// FetchMoveTaskOrder retrieves a MoveTaskOrder for a given UUID
+func FetchMoveTaskOrder(appCtx appcontext.AppContext, searchParams *services.MoveTaskOrderFetcherParams) (*models.Move, error) {
+	mto := &models.Move{}
+
+	query := appCtx.DB().EagerPreload(
+		"Contractor",
```
Comment (ronaktruss) [inline pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go:21]: This is the main line that is different between this and the v1 file

File: pkg/services/move_task_order/shared/utils.go
Comment (ronaktruss) [inline pkg/services/move_task_order/shared/utils.go:1]: We had to refactor some of the existing private functions so that they could be used by either the v1 or v2 version of the service. If you have a better name for this file please let me know.

File: scripts/ensure-spectral-lint [hunk 5]
```diff
@@ -69,7 +69,7 @@ for fn in files:
     # set a really high number that will come down
     unknown_baseline_count = 1024
     baseline_counts = {}
-    if not os.path.exists(baseline_dir):
+    if not os.path.exists(baseline_file):
```
Comment (ronaktruss) [inline scripts/ensure-spectral-lint:72]: This seems like it was a typo. We were getting an error when we added a new file and were trying to find that file.

File: swagger-def/prime_v2.yaml [hunk 6]
```diff
@@ -0,0 +1,1322 @@
+swagger: '2.0'
+info:
+  title: MilMove Prime API
+  version: 0.0.1
```
Comment (ronaktruss) [inline swagger-def/prime_v2.yaml:4]: I think we'll want to update this to 2.0.0 and move our existing APIs to 1.0.0

**Comment by ahobson** [comment]
IMHO, it is a feature to copy all of the yaml object definitions for v2. Otherwise, when you have to make a change to those files, you have to work pretty hard to figure out which versions are affected and which APIs are affected.

IMHO, the prime API should not be using the same object definitions as our other APIs, as we cannot change the prime object definitions, but we can change the ones for our internal APIs.

File: pkg/handlers/routing/api_version/api_version.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+type Flag byte
+
+const (
+	NoneSpecified Flag = iota
+	PrimeVersion1
+	PrimeVersion2
+)
+
+func DetermineAPIVersion(path string) Flag {
```
Comment (kctruss) [inline pkg/handlers/routing/api_version/api_version.go:15]: **thought:** ‏Looking ahead to v3 for example, I'm thinking another way of achieving this would be to explicitly set each version const value (e.g. `PrimeVersion1  Flag = 1` etc) and then parse out the value from the URL to determine which flag is returned. Then the only future update needed to this file would be to add the next incremented version constant.

File: pkg/appcontext/context.go [hunk 8]
```diff
@@ -71,3 +78,11 @@ func (ac *appContext) NewTransaction(fn func(appCtx AppContext) error) error {
...
 	return ac.session
 }
+
+func (ac *appContext) SetAPIVersion(flag apiversion.Flag) {
+	ac.apiVersion = &flag
+}
+
+func (ac *appContext) GetAPIVersion() *apiversion.Flag {
+	return ac.apiVersion
+}
```
Comment (ahobson) [inline pkg/appcontext/context.go:88]: This is pretty different from what I had in mind. Can you help me understand your thought process for adding this to the app context? To me, that is a surprising addition

**Comment by ronaktruss** [comment]
@ahobson Thanks for sharing this feedback! Fwiw, I felt this way initially. Implementing these changes made me change my mind. It felt like we were potentially front-loading work without clear concrete wins.

Can you help me better understand:
> Otherwise, when you have to make a change to those files, you have to work pretty hard to figure out which versions are affected and which APIs are affected.

Would you be able to provide some toy example(s) to demonstrate what you mean? I'm happy to chat if that feels like a more efficient medium. 

Take this with a grain of salt since I'm not sure I am completely following: I think you can tell which API version files are affected by looking at the generated swagger files when you make those changes. From my current vantage point, it seems like both approaches have the same difficulty in determining which API's are affected. 

> IMHO, the prime API should not be using the same object definitions as our other APIs, as we cannot change the prime object definitions without doing a version bump, but we can change the ones for our internal APIs.

This is true. I think we can still ensure that the object definitions are not changing by ensuring that the generated prime yaml file doesn't change. If that is the case, I don't see the advantage of preemptively creating new version of things like PaymentServiceItemStatus (which is just an enum of statuses) and Error when these are things that are unlikely to change and similar across APIs.

That said, there are definitely some downsides to this approach. For example, if you make a change to an object that is only found within another object like address, you'll have to create new v2 definitions of all of the parent objects until you reach the root.

File: pkg/services/move_task_order/move_task_order_v1/move_task_order_fetcher.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
+package movetaskorderv1
```
Comment (ahobson) [inline pkg/services/move_task_order/move_task_order_v1/move_task_order_fetcher.go:1]: I would suggest not using `v1` or `v2` for internal services. Instead, describe the changes. E.g. create a service for `FetchMoveTaskOrderWithDateFilters` or somesuch and use that internally

File: pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go (hunk 4 above)
Comment (ahobson) [reply pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go:21]: How does adding a contractor to `EagerPreload` break the service compatibility? My uninformed guess is that it would preload the `Contractor`, but if the consumer doesn't reference the contractor, it wouldn't matter either way

File: pkg/appcontext/context.go (hunk 8 above)
Comment (ronaktruss) [reply pkg/appcontext/context.go:88]: Some preamble: @mmayo-truss May be able to better chime in since this is pretty much a copy from the work done as part of the ADR. It was mentioned that this was an area that we wanted to look into to see if there was a better way to do so, but because of time considerations this wasn't something that we got to focus in on.

At a high level I believe we wanted a way for the service to dynamically call different versions of the logic so that we wouldn't have to create multiple interfaces for services. One way of doing this was to set a flag and base the service that we are calling on that. And we needed this information while we still had access to the request to determine if it was a v1 or v2 call.

File: pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go (hunk 4 above)
Comment (ronaktruss) [reply pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go:21]: Hmm, I probably should have called this out in the description — I'll add this somewhere in the intro. You are correct, this is a non-breaking change. Ryan had suggested that we make this a major change so that we get a "dry-run" of making a v2 change and that the consumers also have a "dry-run" at making the changes on their side.

Comment (ahobson) [reply pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go:21]: Also: I like the idea of practicing! I feel like there are multiple scenarios to practice:

1. Changing the shape of the data accepted or returned, but keeping the business logic. In that case, I wouldn't expect the services to change.
2. Changing the business logic. Like maybe not returning all the data that used to be returned or something like that. Or for mutations, adding additional restrictions. In this case, the shape of the data remains the same. I think that would mean lots of copying like you did for this, but you could reuse the model to payload logic.
3. Changing both. 

I think it would be interesting to know which scenario you want to try out with this change.  I *think* it's the first one.

File: pkg/appcontext/context.go (hunk 8 above)
Comment (ahobson) [reply pkg/appcontext/context.go:88]: For me, the service is the wrong layer for this. The versioning is at the handler layer, not at the service layer. The service layer is business logic that is independent of version and should be providing business logic operations that are independent of versions. There is a relationship between versions and business logic, and that's what the handler encapsulates.

File: pkg/services/move_task_order/move_task_order_fetcher.go (hunk 2 above)
Comment (ahobson) [reply pkg/services/move_task_order/move_task_order_fetcher.go:101]: If I change this function to only call the v2 service, all the server tests still pass. That suggests to me that for this synthetic scenario, you don't need a new service.

File: pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go (hunk 4 above)
Comment (ronaktruss) [reply pkg/services/move_task_order/move_task_order_v2/move_task_order_fetcher.go:21]: Yeah I think we tried to aim for something closer to option 3 by demonstrating what files might theoretically change. Depending on how loosely we are using business logic, there are some instances where we'd update the handler to call an entirely new service, just update the service, or just update the model_to_payload/payload_to_model changes if we wanted to add or remove fields. Hopefully this PR demonstrates what those changes might look like either explicitly or by extrapolating the current changes to those needed. It doesn't cover business logic changes like modifying validations.

File: pkg/services/move_task_order/move_task_order_fetcher.go (hunk 2 above)
Comment (ronaktruss) [reply pkg/services/move_task_order/move_task_order_fetcher.go:101]: Yeah we don't necessarily need them. They were included primarily to demonstrate what those changes would potentially look like.

File: pkg/appcontext/context.go (hunk 8 above)
Comment (mmayo-truss) [reply pkg/appcontext/context.go:88]: The option chosen for versioning services came out of [this conversation](https://github.com/transcom/mymove-docs/pull/351#discussion_r1223755497). It was an attempt to organize different versions of our services when the need arose. 

What you are saying is that we wouldn't need to version our services at all? Instead, if needed, we could create a different service and call that with the new version's handler. This seems more like [the first option](https://github.com/transcom/mymove/pull/10816/commits/0913f2f4625450a2ac43b358e1890d17d77b747d) considered for services in the ADR. Though instead of renaming the old service, you would make the new service name something more descriptive.

Comment (ahobson) [reply pkg/appcontext/context.go:88]: Yes, I apologize. I believe I missed some details in the ADR.

I think it depends on the modifications needed, but I believe frequently you may be able to keep the same service and add a new method on the interface. Then you can, if feasible, refactor to share code between the two methods as needed.

We may need to get some experience to see what kinds of changes require service modifications. I think trying to plan it out too far in advance of actual experience can be a dangerous thing. (Not that I've ever done that. No. Never. Me? No.)

File: pkg/handlers/routing/routing_init.go [hunk 10]
```diff
@@ -252,16 +255,21 @@ func InitRouting(appCtx appcontext.AppContext, redisPool *redis.Pool,
 		primeMux.Use(authentication.PrimeAuthorizationMiddleware(appCtx.Logger()))
 		primeMux.Use(middleware.NoCache(appCtx.Logger()))
 		primeMux.Use(middleware.RequestLogger(appCtx.Logger()))
-		primeMux.HandleFunc("/swagger.yaml", handlers.NewFileHandler(routingConfig.PrimeSwaggerPath)).Methods("GET")
+		primeMux.HandleFunc("/v1/swagger.yaml", handlers.NewFileHandler(routingConfig.PrimeSwaggerPath)).Methods("GET")
+		primeMux.HandleFunc("/v2/swagger.yaml", handlers.NewFileHandler(routingConfig.PrimeV2SwaggerPath)).Methods("GET")
 		if routingConfig.ServeSwaggerUI {
 			appCtx.Logger().Info("Prime API Swagger UI serving is enabled")
 			primeMux.HandleFunc("/docs", handlers.NewFileHandler(path.Join(routingConfig.BuildRoot, "swagger-ui", "prime.html"))).Methods("GET")
+			primeMux.HandleFunc("/docs", handlers.NewFileHandler(path.Join(routingConfig.BuildRoot, "swagger-ui", "prime_v2.html"))).Methods("GET")
```
Comment (ronaktruss) [inline pkg/handlers/routing/routing_init.go:263]: To enable supporting this we needed to modify the public/swagger-ui folder and the gitignore file. I don't think folks generally use the swagger UI anymore now that we have the redocly api documentation, but it wasn't a large lift to continue to support this.

File: swagger-def/prime_v2.yaml (hunk 6 above)
Comment (kctruss) [reply swagger-def/prime_v2.yaml:4]: This makes sense to me.

**Comment by duncan-truss** [comment]
@ronaktruss I think we'll need to add prime2 manually to our spectral linter list in our .pre-commit-config.yaml  https://github.com/transcom/mymove/blob/4098258c65e383e3540732dc2a23ee66c3f3bf9d/.pre-commit-config.yaml#L182

File: pkg/handlers/routing/routing_init.go [hunk 11]
```diff
@@ -252,16 +255,21 @@ func InitRouting(appCtx appcontext.AppContext, redisPool *redis.Pool,
 		primeMux.Use(authentication.PrimeAuthorizationMiddleware(appCtx.Logger()))
 		primeMux.Use(middleware.NoCache(appCtx.Logger()))
 		primeMux.Use(middleware.RequestLogger(appCtx.Logger()))
-		primeMux.HandleFunc("/swagger.yaml", handlers.NewFileHandler(routingConfig.PrimeSwaggerPath)).Methods("GET")
+		primeMux.HandleFunc("/v1/swagger.yaml", handlers.NewFileHandler(routingConfig.PrimeSwaggerPath)).Methods("GET")
+		primeMux.HandleFunc("/v2/swagger.yaml", handlers.NewFileHandler(routingConfig.PrimeV2SwaggerPath)).Methods("GET")
```
Comment (ahobson) [inline pkg/handlers/routing/routing_init.go:259]: In the old setup, the path prefix was `/prime/v1` with `/swagger.yaml` served under it, `/docs` optionally served, and then the API mounted under it.

I'd suggest following the same pattern with `/prime/v2` instead of trying to combine them. Maybe rename `primeMux` to `primeV1Mux` and then created a `primeV2Mux`. Below, I believe you have two routes for `/docs` that conflict with each other.

**Review by ahobson — COMMENTED** [review]
I personally like this approach a LOT! I think it is easier to follow.

File: pkg/handlers/routing/routing_init.go (hunk 11 above)
Comment (mmayo-truss) [reply pkg/handlers/routing/routing_init.go:259]: Running this locally, I am able to see both the prime v1 and the prime v2 swagger-ui pages at `http://primelocal:3000/swagger-ui/prime.html` and `http://primelocal:3000/swagger-ui/prime_v2.html` respectively. Perhaps each `/docs` route could be differentiated?

Having a separate mux for each version would mean reapplying all of the same middleware. I can see this getting cumbersome, especially if more major versions are ever introduced.

Comment (ahobson) [reply pkg/handlers/routing/routing_init.go:259]: I dunno, maybe? I mean, how often do we anticipate adding another major version? Every six months or so?

I would personally prefer to be explicit with some copy and paste

Comment (ahobson) [reply pkg/handlers/routing/routing_init.go:259]: I might be wrong!

We could add the middleware on e.g. the `/prime` route and then mount `/v1`/ and `/v2` under it?

Comment (ronaktruss) [reply pkg/handlers/routing/routing_init.go]: Yeah, I was just taking a look at doing something like that since it looks like thats what we are doing for the subrouters from the main `site` router. I believe this works as expected. I'm not too familiar with mux, but this seems like a valid pattern after reading through some other examples and docs.

The changes can be seen [here](https://github.com/transcom/mymove/pull/10952/files/b9df3c26fafc670403b8f03dc69db01814ab89ca..3ca97ceda4a44c4d8a2d5abb19cbc1bfba0691a1).

File: swagger-def/prime_v2.yaml [hunk 12]
```diff
@@ -0,0 +1,82 @@
+swagger: '2.0'
+info:
+  title: MilMove Prime API
```
Comment (duncan-truss) [inline swagger-def/prime_v2.yaml:3]: This does get used in the `<title>` tag I think so might be nice to distinguish for folks bookmarking it but it should be pretty obvious based on the rest of the info.

**Comment by duncan-truss** [comment]
I don't think it's a blocker but our `cmd/prime-api-client` is only capable of calling the existing v1 endpoints I think.  Maybe we can document that in the usage notes somewhere so the limitation is obvious.

The other thing I wanted to ask was about the audit log event name.  It seems to default to the operationID of the endpoint.  I wonder if we need a convention to distinguish between events that originate in v1 from v2 https://github.com/transcom/mymove/blob/d2064d4d9fa9f223731beb089420b36035533c87/pkg/audit/context.go#L11.  Seems like it will be more important for mutations.

File: swagger-def/prime_v2.yaml (hunk 12 above)
Comment (ronaktruss) [reply swagger-def/prime_v2.yaml]: Updated. I'll also call out that there is another ticket that will be used to actually make this yaml file available in our docs.

**Comment by ronaktruss** [comment]
> I don't think it's a blocker but our cmd/prime-api-client is only capable of calling the existing v1 endpoints I think. Maybe we can document that in the usage notes somewhere so the limitation is obvious.

@duncan-truss Yeah that's a good call out. There is some overlap between leveraging postman to call prime/support apis, the prime api client, and the prime simulator. I don't think the boundaries between these are very clear and we'll likely need alignment on what we are actually planning on supporting long term. If you think otherwise feel free to let me know.

I don't have much context on the audit logs. Let me try and do some digging.

**Comment by duncan-truss** [comment]
@ronaktruss Should the prime simulator be updated to call v2 of getMoveTaskOrder?  I know we only have one swagger client setup and would need to add one for v2 if we want to have both availaible. https://github.com/transcom/mymove/blob/1fc9901379cc8ec5921f090e7b0e662ec45cd06e/src/services/primeApi.js#L8-L17

**Comment by ronaktruss** [comment]
@duncan-truss Right now the prime simulator wasn't updated to be able to support this new API call. This was intentional for a couple of reasons: 
1. It's unclear to me how we are supporting v2 APIs if they are not being adopted immediately. In the case we introduce a v2 API but the prime is still using a v1 API, would we want to update the prime simulator to just default to the new version? I think the answer depends on the specifics of the scenario. 
2. To limit the scope of the this initial PR. Because the answer to the question above feels a bit muddy it didn't seem like it was a necessary inclusion at the time. 

That said, we could still add the capability of accessing the v2 APIs without changing the existing behaviors. I was planning on doing this in a separate PR after getting a bit more clarity on how this PR would land. If you think that should be included in this PR feel free to let me know. I can be easily swayed on this.

**Comment by ronaktruss** [comment]
>The other thing I wanted to ask was about the audit log event name. It seems to default to the operationID of the endpoint. I wonder if we need a convention to distinguish between events that originate in v1 from v2. Seems like it will be more important for mutations. 

Also I spoke to @mmayo-truss about this and it seems unlikely that we will want to distinguish between a v1 and v2 event. (I think he might be writing a response that will likely have more specifics.) From the UI perspective the user likely doesn't care between the distinction between the two. There might be some edge cases where entities are being significantly modified but those will likely be paired with their own set of template changes.

**Comment by mmayo-truss** [comment]
> I don't think it's a blocker but our `cmd/prime-api-client` is only capable of calling the existing v1 endpoints I think. Maybe we can document that in the usage notes somewhere so the limitation is obvious.
> 
> The other thing I wanted to ask was about the audit log event name. It seems to default to the operationID of the endpoint. I wonder if we need a convention to distinguish between events that originate in v1 from v2. Seems like it will be more important for mutations.
> 
> https://github.com/transcom/mymove/blob/d2064d4d9fa9f223731beb089420b36035533c87/pkg/audit/context.go#L11

@duncan-truss The event name is used in conjunction with the action type and the table name to select the appropriate template to display in the for a row in audit history table UI. If a unique combination of those three are not found a default template would be utilized. [This is the template registration function](https://github.com/transcom/mymove/blob/e3fa8740225cc9e5f4320cbb7835ed8a5f6e13fb/src/constants/MoveHistory/TemplateManager.js#L16-L18). [And this is an example template for an update to MTO Service Item](https://github.com/transcom/mymove/blob/f63ca59a1f96c967afccf8e366638f4d276cd70b/src/constants/MoveHistory/EventTemplates/UpdateMTOServiceItem/updateMTOServiceItem.jsx#L20-L27).

If we differentiated v1 and v2 event names here, we would need to have separate templates in the audit history table for each. Or modify the template lookup mechanism to account for the multiple versions of the event name.

I think defaulting to using the same event name makes sense. If, in the process of versioning an endpoint, the need to differentiate the templates for the audit history table UI arises, we could differentiate the one endpoint and create the necessary audit history event templates.

File: swagger-def/prime_v2.yaml (hunk 12 above)
Comment (duncan-truss) [reply swagger-def/prime_v2.yaml]: Thank you, I think this made things clearer in my Postman import too

**Review by duncan-truss — APPROVED** [review]
Thanks for indulging my questions and helping me understand what we might do in the future.  Verified it Postman and am only seeing a few unused lint warnings.

**Comment by ronaktruss** [comment]
@duncan-truss  Are you seeing any warnings outside `oas2-unused-definition`? I think it's expected that there should be 5 warnings atm.

**Comment by duncan-truss** [comment]
> @duncan-truss Are you seeing any warnings outside `oas2-unused-definition`? I think it's expected that there should be 5 warnings atm.

Those were the only ones I saw locally and I think that count matched what was in CI too.

**Review by LeDeep — APPROVED** [review]

File: pkg/handlers/routing/routing_init.go [hunk 13]
```diff
@@ -252,16 +256,29 @@ func InitRouting(appCtx appcontext.AppContext, redisPool *redis.Pool,
 		primeMux.Use(authentication.PrimeAuthorizationMiddleware(appCtx.Logger()))
 		primeMux.Use(middleware.NoCache(appCtx.Logger()))
 		primeMux.Use(middleware.RequestLogger(appCtx.Logger()))
-		primeMux.HandleFunc("/swagger.yaml", handlers.NewFileHandler(routingConfig.PrimeSwaggerPath)).Methods("GET")
+
+		// Setup version specific info
```
Comment (mmayo-truss) [inline pkg/handlers/routing/routing_init.go:260]: Praise: I like the use of the subrouters here after the prime middleware was applied.

**Review by mmayo-truss — APPROVED** [review]
Good work on this. This was a lot of work and coordination. I also appreciate the thoughts you gave on the process. This will help us version other endpoints going forward.
